clang/test/CheckedC/inferred-bounds/widened-bounds.c

```diff
@@ -464,3 +466,72 @@ void f19() {
 // CHECK:  [B1]
 // CHECK: upper_bound(p) = 3
 }
+
+void f20() {
+  // Declared bounds and deref offset INT_MAX. Valid widenening.
+  _Nt_array_ptr<char> p : count(INT_MAX) = "";      // expected-error {{declared bounds for 'p' are invalid after initialization}}
+  if (*(p + INT_MAX))
+  {}
+
+// CHECK: In function: f20
+// CHECK:  [B12]
+// CHECK:    1: _Nt_array_ptr<char> p : count(2147483647) = "";
+// CHECK:    2: *(p + 2147483647)
+// CHECK:  [B11]
+// CHECK: upper_bound(p) = 1
+
+  // Declared bounds and deref offset INT_MIN. Valid widenening.
+  _Nt_array_ptr<char> q : count(INT_MIN) = "";
+  if (*(q + INT_MIN))                               // expected-error {{out-of-bounds memory access}}
+  {}
+
+// CHECK:  [B10]
+// CHECK:    1: _Nt_array_ptr<char> q : count((-2147483647 - 1)) = "";
+// CHECK:    2: *(q + (-2147483647 - 1))
+// CHECK:  [B9]
+// CHECK: upper_bound(q) = 1
+
+  // Declared bounds (INT_MIN) and deref offset (INT_MAX). No sequential deref tests. No widenening.
+  _Nt_array_ptr<char> r : count(INT_MIN) = "";
+  if (*(r + INT_MAX))                               // expected-error {{out-of-bounds memory access}}
+  {}
+
+// CHECK:  [B8]
+// CHECK:    1: _Nt_array_ptr<char> r : count((-2147483647 - 1)) = "";
+// CHECK:    2: *(r + 2147483647)
+// CHECK:  [B7]
+// CHECK-NOT: upper_bound(r)
+
+  // Declared bounds and deref offset (INT_MAX + 1). Integer overflow. No widenening.
```

Comment:
There a typo in the comment and a few other comments: `widenening` -> `widening`.